include/oneapi/dpl/pstl/hetero/algorithm_impl_hetero.h

```diff
@@ -1198,17 +1205,25 @@ oneapi::dpl::__internal::__enable_if_hetero_execution_policy<_ExecutionPolicy, v
 __pattern_stable_sort(_ExecutionPolicy&& __exec, _Iterator __first, _Iterator __last, _Compare __comp,
                       /*vector=*/::std::true_type, /*parallel=*/::std::true_type)
 {
-    if (__last - __first < 2)
-        return;
+    __stable_sort_pattern_impl(::std::forward<_ExecutionPolicy>(__exec), __first, __last, __comp, oneapi::dpl::identity{});
+}
 
-    auto __keep = oneapi::dpl::__ranges::__get_sycl_range<__par_backend_hetero::access_mode::read_write, _Iterator>();
-    auto __buf = __keep(__first, __last);
+template <typename _ExecutionPolicy, typename _Iterator1, typename _Iterator2, typename _Compare>
+oneapi::dpl::__internal::__enable_if_hetero_execution_policy<_ExecutionPolicy, void>
+__pattern_sort_by_key(_ExecutionPolicy&& __exec, _Iterator1 __keys_first, _Iterator1 __keys_last,
+                      _Iterator2 __values_first, _Compare __comp, /*vector=*/::std::true_type,
+                      /*parallel=*/::std::true_type)
+{
+    static_assert(::std::is_move_constructible_v<_Iterator1> && ::std::is_move_constructible_v<_Iterator2>,
+        "The keys abd values should be move constructible in case of parallel execution.");
```

Comment:
Same thing here with the `::std::is_move_constructible_v` for the iterator instead of its value type.